### PR TITLE
Connection UI: Building the Framework

### PR DESCRIPTION
**Pull request by sergeymitr**
#### Changes proposed in this Pull Request:
- The new package
- JS building configuration
- Handy yarn commands
- The WP admin menu link

The new commands (run from `packages/connection-ui/` directory):
- `yarn build`
- `yarn watch`

The Jetpack command `yarn build-packages` was also edited to also build the Connection UI package.

#### Jetpack product discussion
p9dueE-1vM-p2
Related to #18182

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Go to `packages/connection-ui` and run `yarn install` to install the required modules.
2. Run `yarn build`, confirm no errors come up.
3. Go to WP admin area, "Tools -> Connection Manager", you should see the "Connection Manager" header. That means the build was successful.
4. Run `yarn watch`.
5. Open the file `packages/connection-ui/_inc/admin.jsx` and change the `Connection Manager` text to something else.
6. Wait a few seconds, reload the "Tools -> Connection Manager" page, make sure the changed text appears.
7. Stop the watching process (Ctrl+C).
8. Revert the file `packages/connection-ui/_inc/admin.jsx` to its original state.
9. Go to the Jetpack root directory and run `yarn build-packages`. Confirm that it shows no errors.
10. Once finished, reload the "Tools -> Connection Manager" page again. Make sure it reflects the latest changes.

_The PR uses a deprecated function `registerStore()` of the `@wordpress/data` component.
Unfortunately, we can't replace it with `createReduxStore()` and `register()` right now, because they are not included into the latest Wordpress build of `@wordpress/data`._

#### Proposed changelog entry for your changes:
n/a

### CODE REVIEW

**Comment by kraftbj** [comment]
As an FYI, later this week, GC will be merging our repo reorg patch that, will among other things, allow for individual projects--like this new package--to build and commit to their mirror repo a JS-built version.

In other words, Composer clients can consume the package and get the built, ready to use JS after we merge and set things up for this package.

File: packages/connection-ui/package.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+		"webpack": "4.45.0"
+	},
+	"devDependencies": {
+		"@wordpress/dependency-extraction-webpack-plugin": "2.9.0",
+		"enzyme": "^3.11.0",
+		"jest": "26.6.3"
+	},
+	"engines": {
+		"node": "^12.19.1",
+		"yarn": "^1.22.10"
```
Comment (jeherve) [inline packages/connection-ui/package.json:37]: Should we stick to dependencies similar to the one in the main repo, to avoid having to switch between node versions when working in one vs. the other?

```suggestion
		"node": "^12.18.0",
		"yarn": "^1.3.2"
```

**Review by jeherve — CHANGES_REQUESTED** [review]

File: packages/connection-ui/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+const path = require( 'path' );
+
+const isDevelopment = process.env.NODE_ENV !== 'production';
+
+const baseConfig = getBaseWebpackConfig(
+	{ WP: false },
+	{
+		entry: {}, // We'll override later
+		'output-filename': '[name].js',
+		'output-path': path.join( __dirname, './build' ),
```
Comment (zinigor) [inline packages/connection-ui/webpack.config.js:15]: I guess this is as good as a place to discuss as any, but what do you think about how easy/hard it would be for us to use that package? I mean, the consuming package (or if we decide to combine it with the connection package, the package itself) will need to use it. We're already doing some enqueuing, asset loading, etc. Should we be putting the built files inside the package itself, or should we be using the built Jetpack folder?
Basically what I'm asking is **What would a release of this package look like**?

**Review by zinigor — CHANGES_REQUESTED** [review]
This is a great start, thank you! We're moving in the right direction, I have left some comments in the code, please take a look.

File: packages/connection-ui/package.json (hunk 1 above)
Comment (sergeymitr) [reply packages/connection-ui/package.json]: That's a really good point, thanks!

File: composer.json [hunk 3]
```diff
@@ -20,6 +20,7 @@
 		"automattic/jetpack-compat": "@dev",
 		"automattic/jetpack-config": "@dev",
 		"automattic/jetpack-connection": "@dev",
+		"automattic/jetpack-connection-ui": "@dev",
```
[the remark this answers is not in the document]
Comment (sergeymitr) [reply composer.json:23]: Hi @zinigor.

I see following benefits in having the `connection-ui` as a separate package:
1. Separation of responsibilities. The UI is a thing in itself, and although it relies on the `connection` package, they are loosely coupled. So it looks like a logical path to keep developing them separately, as well as tag new versions of each separately.
2. I can imagine a scenario where `connection` package would be used without `connection-ui`. For example, if a plugin tries not to follow its own branding and avoid having Jetpack and WP.com mentioned anywhere in its functionality. For example, WooCommerce Payments have had its connection process adjusted for that purpose.
3. I'm a huge proponent of the "building blocks" approach of Unix systems. If we keep packages slim, plugins will be able to choose the exact they need, and skip the rest. I feel like the `connection` package is already in need of refactoring (e.g. the `jetpack-api` package initiative), and making it even bigger will make its maintenance more taxing.

File: packages/connection-ui/webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * External dependencies
+ */
+const getBaseWebpackConfig = require( '@automattic/calypso-build/webpack.config.js' );
+const DependencyExtractionWebpackPlugin = require( '@wordpress/dependency-extraction-webpack-plugin' );
+const path = require( 'path' );
+
+const isDevelopment = process.env.NODE_ENV !== 'production';
+
+const baseConfig = getBaseWebpackConfig(
+	{ WP: false },
```
[the remark this answers is not in the document]
Comment (sergeymitr) [reply packages/connection-ui/webpack.config.js:11]: The `env.WP` flag does two things:
1. Initializes the `DependencyExtractionWebpackPlugin` to not include the dependencies available in WordPress into the build. We do that separately by including the plugin manually in that config.
2. Creates an abstraction layer using `@wordpress/element`. I don't see any potential need for it, so I think it would be an excessive complexity. We can still use the `@wordpress/element` components (e.g. `Button`) when needed.

This is the way we handle things in Jetpack, this logic is copied from there.

File: packages/connection-ui/webpack.config.js (hunk 2 above)
Comment (sergeymitr) [reply packages/connection-ui/webpack.config.js:15]: There are going to be two files in the `./build` directory: `index.js` and `index.css`.
The plan is to load them both when Connection Manager UI page is requested: https://github.com/Automattic/jetpack/pull/18303/files#diff-e56f98a9a60baeb985dc0456cfe3c301ed65bc05c050d1c62250bbc589487d19R56-R61

> should we be using the built Jetpack folder

I'm not sure I understand the idea, could you elaborate.

> What would a release of this package look like?

The release process wouldn't change a bit. The package is getting build by the existing command `yarn build-packages`, which in turn is a part of the `yarn build`.
The package will have its own `./build` directory, and its own `enqueue` calls in the PHP code to load them on demand.
The only downside of this approach I see is that each plugin will have its own copy of the built JS and CSS files, but they aren't going to be too big to worry about the size.

Plugins other than Jetpack will need to run `yarn --cwd ./packages/connection-ui build` to build the package, or somehow integrate it into their own build system.

I'm also wrestling with an idea to remove the `./_inc` directory with the actual JS source code while building a package for a release. That would make the package slimmer.
A command like `yarn --cwd ./packages/connection-ui build-release` would run the regular build, and run the `rm -rf _inc` afterwards.

Please let me know what you think, as I might be missing some underlying issues.

Comment (kraftbj) [reply packages/connection-ui/webpack.config.js:15]: To add on to https://github.com/Automattic/jetpack/pull/18303#issuecomment-758347165 , where GC is going is that after a PR is merged to `master` that changed a package file, in addition to committing the change to the mirror repo, it will also run a build step (see https://github.com/Automattic/jetpack/pull/18303#issuecomment-758347165 for where it's at now, but it'll change as we have more than Jetpack the plugin using it).

The idea here in Composer, a consuming project shouldn't need to do any building of JS.

I'm open to other ideas and this is something that we added in as part of the reorg while I was working on getting a built version of Jetpack copied over to the new jetpack-production repo.

Comment (jeherve) [reply packages/connection-ui/webpack.config.js:15]: We've had a somehow similar discussion in #17571 and #17562, but went in a different direction. It may be good to agree on a common solution for all our packages.

> Plugins other than Jetpack will need to run yarn --cwd ./packages/connection-ui build to build the package, or somehow integrate it into their own build system.

I think this may be problematic for some. I think part of the assumptions behind PHP packages is that just adding them as a dependency for your project gives you access to everything they have to offer, without having to run commands afterwards to edit the contents of the `vendor/` directory in your app.

Comment (sergeymitr) [reply packages/connection-ui/webpack.config.js:15]: Hi Jeremy, Kraft.

These are very good points. Let's wait for the aforementioned repo reorganizing to be merged then, and go from there.

File: composer.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (zinigor) [reply composer.json:23]: > Separation of responsibilities. The UI is a thing in itself, and although it relies on the connection package, they are loosely coupled. So it looks like a logical path to keep developing them separately, as well as tag new versions of each separately.

I agree that in a package that would only have the client side things it would be better to keep things separate, but the UI will definitely need additional things on the server side too. Something will have to provide the API that feeds the data to the client side, as we have seen in the proof-of-concept PR. From here I would argue that it blurs the line between UI and connection itself.

> I can imagine a scenario where connection package would be used without connection-ui. For example, if a plugin tries not to follow its own branding and avoid having Jetpack and WP.com mentioned anywhere in its functionality. For example, WooCommerce Payments have had its connection process adjusted for that purpose.

Would the benefit of not shipping UI in this case be substantial? Currently we are only planning to ship several elements. Although you may have a point in case of the RNA admin interface.

> I'm a huge proponent of the "building blocks" approach of Unix systems. If we keep packages slim, plugins will be able to choose the exact they need, and skip the rest. I feel like the connection package is already in need of refactoring (e.g. the jetpack-api package initiative), and making it even bigger will make its maintenance more taxing.

Yes, I do see your point - I'm trying to find arguments against separating the packages, and the biggest one so far is the first thing I mentioned - needing to find the line where to separate the UI from the rest of the stuff. 
OK, let's go with a separate package for now and we can see about it later, because we're now mostly concerned about the build part.

File: package.json [hunk 5]
```diff
@@ -20,7 +20,7 @@
 		"build-color-schemes-wpcom": "GULP_ENV=wpcom gulp sass:color-schemes",
 		"build-concurrently": "yarn install-if-deps-outdated && yarn clean && yarn concurrently 'yarn build-client' 'yarn build-php' 'yarn build-extensions' 'yarn build-search' 'yarn build-packages'",
 		"build-extensions": "webpack --config ./webpack.config.extensions.js",
-		"build-packages": "webpack --config ./webpack.config.packages.js",
+		"build-packages": "webpack --config ./webpack.config.packages.js && yarn --cwd ./packages/connection-ui build",
```
Comment (zinigor) [inline package.json:23]: Now that we have merged the reorg PR, can we go about building JS for this package the same way we do it for lazy images? I mean two things here:

1. Putting build files back into the package folder, and enqueuing them from there.
2. Using the same type of script that we use for lazy images in the package.json.

File: packages/connection-ui/webpack.config.js (hunk 2 above)
Comment (sergeymitr) [reply packages/connection-ui/webpack.config.js]: So the repo reorg has been merged, and I rebased the PR to include them.

> our repo reorg patch that, will among other things, allow for individual projects--like this new package--to build and commit to their mirror repo a JS-built version.

That seems to be the ideal scenario, publishing a built version into the package repo, so it could be included into other plugins without any need for building the JS.
@kraftbj, is there any documentation on publishing pre-built packages I could review?

File: package.json (hunk 5 above)
Comment (sergeymitr) [reply package.json]: We're doing it essentially the same way.
The only difference is that we use Gulp for this package, while Lazy Images setup is a bit simpler and doesn't need it.

**Comment by sergeymitr** [comment]
Rebased to resolve a conflict with the `master` branch.

**Comment by sergeymitr** [comment]
Hi @jeherve, @kraftbj, @zinigor.
Can you please give the PR a final look, as believe it's ready to be merged.

File: projects/packages/connection-ui/composer.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+	"require": {
+		"automattic/jetpack-connection": "@dev"
+	},
+	"autoload": {
+		"classmap": [
+			"src/"
+		]
+	},
+	"repositories": [
+		{
```
Comment (kraftbj) [inline projects/packages/connection-ui/composer.json:15]: Please tag this per the error above.

File: projects/packages/connection-ui/composer.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+			"type": "path",
+			"url": "../*"
+		}
+	],
+	"minimum-stability": "dev",
```
Comment (kraftbj) [inline projects/packages/connection-ui/composer.json:20]:
```suggestion
	"scripts": {
		"build-development": [
			"Composer\\Config::disableProcessTimeout",
			"yarn build"
		]
	},
	"minimum-stability": "dev",
```

There's a new `jetpack build` command that just merged that would need this to build the package centrally.

File: projects/packages/connection-ui/package.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+	"description": "Jetpack Connection Manager UI",
+	"main": "_inc/admin.jsx",
+	"repository": "https://github.com/Automattic/jetpack-connection-ui",
+	"author": "Automattic",
+	"license": "GPL-2.0-or-later",
+	"scripts": {
+		"build": "yarn install-if-deps-outdated && yarn clean && yarn build-client",
+		"build-client": "gulp",
+		"clean": "rm -rf build/",
+		"install-if-deps-outdated": "yarn check 2> /dev/null || yarn install --check-files --production=false --frozen-lockfile",
```
Comment (kraftbj) [inline projects/packages/connection-ui/package.json:13]:
```suggestion
		"install-if-deps-outdated": "yarn install --check-files --production=false --frozen-lockfile",
```

We can remove this. It was removed upstream from the main one, it's a deprecated check that doesn't do anything different than yarn install --check-files.

**Review by kraftbj — CHANGES_REQUESTED** [review]
A few monorepo-related things. Didn't look over the actual functionality :)

File: packages/connection-ui/webpack.config.js (hunk 2 above)
Comment (kraftbj) [reply packages/connection-ui/webpack.config.js]: Sorry for missing this comment until now! Is there a mirror repo already created for this one? If so, we'd need to add it to the `composer.json`. See lazy-images for an example: https://github.com/Automattic/jetpack/blob/master/projects/packages/lazy-images/composer.json#L45

With a `build-production` and a mirror-repo set, the GH action will automatically build it and put the built version in the mirror repo.

File: projects/packages/connection-ui/composer.json (hunk 7 above)
Comment (sergeymitr) [reply projects/packages/connection-ui/composer.json:33]: Done, thanks!

File: projects/plugins/jetpack/class.jetpack.php [hunk 9]
```diff
@@ -824,6 +824,7 @@ public function configure() {
 
 		if ( ! $this->connection_manager ) {
 			$this->connection_manager = new Connection_Manager( 'jetpack' );
+			Automattic\Jetpack\ConnectionUI\Admin::init();
```
Comment (zinigor) [inline projects/plugins/jetpack/class.jetpack.php:827]: Since this is still a work in progress, what do you say we add a filter to conditionally enable initialization of the Connection UI, so we can disable it by default until it's ready?

**Review by zinigor — CHANGES_REQUESTED** [review]
One minor change, and I think this is ready to be shipped. Thanks for working on this!

File: projects/packages/connection-ui/composer.json [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+			"yarn build"
+		],
+		"build-production": [
+			"Composer\\Config::disableProcessTimeout",
+			"yarn build"
+		]
+	},
+	"minimum-stability": "dev",
+	"prefer-stable": true
+}
```
Comment (jeherve) [inline projects/packages/connection-ui/composer.json:35]: I believe we should add `extra.mirror-repo` here, as per the requirements:
https://github.com/Automattic/jetpack/blob/5c5ead66f8e7e59f93423eb7f34a54b5132daca2/docs/monorepo.md#L46

File: projects/packages/connection-ui/src/class-admin.php [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	/**
+	 * Register's submenu.
+	 */
+	public function register_submenu_page() {
+		add_submenu_page(
+			'tools.php',
+			'Connection Manager',
+			'Connection Manager',
```
Comment (jeherve) [inline projects/packages/connection-ui/src/class-admin.php:42]: You'll probably want those to be translatable:

```suggestion
			__( 'Connection Manager', 'jetpack' ),
			__( 'Connection Manager', 'jetpack' ),
```

**Review by jeherve — CHANGES_REQUESTED** [review]
I think we would need a `.gitattributes` in here to specify what we don't want to ship with the production package.

File: projects/packages/connection-ui/package.json [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+		"static-site-generator-webpack-plugin": "3.4.2",
+		"webpack": "4.45.0"
+	},
+	"devDependencies": {
+		"@wordpress/dependency-extraction-webpack-plugin": "2.9.0",
+		"enzyme": "^3.11.0",
+		"jest": "26.6.3"
+	},
+	"engines": {
+		"node": "^12.18.0",
```
Comment (jeherve) [inline projects/packages/connection-ui/package.json:36]: We've now updated to `12.20.1` for the rest of the monorepo.

```suggestion
		"node": "^12.20.1",
```

**Review by zinigor — DISMISSED** [review]
Let's get this party started!

**Review by kraftbj — APPROVED** [review]

File: projects/packages/connection-ui/.gitattributes [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+gulpfile.babel.js export-ignore
+.babelrc          export-ignore
+
+# Files to include in the mirror repo
+/build/index.asset.php  production-include
+/build/index.js         production-include
+
+# Files to exclude from the mirror repo
+.gitignore        production-exclude
+_inc/             production-exclude
```
Comment (anomiex) [inline projects/packages/connection-ui/.gitattributes:14]: This tags the directory, but not the files inside it. So they still get included.
```suggestion
_inc/**           production-exclude
```

**Review by anomiex — APPROVED** [review]
Adding approval based on the previous version receiving approvals from others. Haven't reviewed myself.

**Review by kraftbj — APPROVED** [review]
